pkg/ruler/compat.go

```diff
@@ -229,19 +229,25 @@ func MetricsQueryFunc(qf rules.QueryFunc, queries, failedQueries prometheus.Coun
 	}
 }
 
-func RecordAndReportRuleQueryMetrics(qf rules.QueryFunc, queryTime prometheus.Counter, logger log.Logger) rules.QueryFunc {
-	if queryTime == nil {
-		return qf
-	}
-
+func RecordAndReportRuleQueryMetrics(qf rules.QueryFunc, userID string, evalMetrics *RuleEvalMetrics, logger log.Logger) rules.QueryFunc {
 	return func(ctx context.Context, qs string, t time.Time) (promql.Vector, error) {
+		queryTime := evalMetrics.RulerQuerySeconds.WithLabelValues(userID)
+		querySeries := evalMetrics.RulerQuerySeries.WithLabelValues(userID)
+		querySample := evalMetrics.RulerQuerySamples.WithLabelValues(userID)
+		queryChunkBytes := evalMetrics.RulerQueryChunkBytes.WithLabelValues(userID)
+		queryDataBytes := evalMetrics.RulerQueryDataBytes.WithLabelValues(userID)
```

Comment:
these should be outside of the scope of this anonymous function. they should be in the scope of `RecordAndReportRuleQueryMetrics`.

Reply:
I have fixed it.